Source/ablastr/math/fft/WrapMklFFT.cpp

```diff
@@ -0,0 +1,84 @@
+/* Copyright 2019-2023
+ *
+ * This file is part of ABLASTR.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+
+#include "AnyFFT.H"
+
+#include "ablastr/utils/TextMsg.H"
+#include "ablastr/profiler/ProfilerWrapper.H"
+
+#include <cstdint>
+
+namespace ablastr::math::anyfft
+{
+
+    void setup(){/*nothing to do*/}
+
+    void cleanup(){/*nothing to do*/}
+
+    FFTplan CreatePlan(const amrex::IntVect& real_size, amrex::Real * const real_array,
+                       Complex * const complex_array, const direction dir, const int dim)
+    {
+        FFTplan fft_plan;
+        ABLASTR_PROFILE("ablastr::math::anyfft::CreatePlan");
+
+        // Initialize fft_plan.m_plan with the vendor fft plan.
+        if (dim == 3) {
+            fft_plan.m_plan = new std::remove_pointer_t<VendorFFTPlan>(
+                {std::int64_t(real_size[2]),
+                 std::int64_t(real_size[1]),
+                 std::int64_t(real_size[0])});
+        } else if (dim == 2) {
+            fft_plan.m_plan = new std::remove_pointer_t<VendorFFTPlan>(
+                {std::int64_t(real_size[1]),
+                 std::int64_t(real_size[0])});
+        } else if (dim == 1) {
+            fft_plan.m_plan = new std::remove_pointer_t<VendorFFTPlan>(
+                std::int64_t(real_size[0]));
+        } else {
+            ABLASTR_ABORT_WITH_MESSAGE("only dim2 =1, dim=2 and dim=3 have been implemented");
+        }
+
+        fft_plan.m_plan->set_value(oneapi::mkl::dft::config_param::PLACEMENT,
+                                   DFTI_NOT_INPLACE);
+        fft_plan.m_plan->commit(amrex::Gpu::Device::streamQueue());
+
+        // Store meta-data in fft_plan
+        fft_plan.m_real_array = real_array;
+        fft_plan.m_complex_array = complex_array;
+        fft_plan.m_dir = dir;
+        fft_plan.m_dim = dim;
+        fft_plan.m_stream = amrex::Gpu::gpuStream();
+
+        return fft_plan;
+    }
+
+    void DestroyPlan(FFTplan& fft_plan)
+    {
+        delete fft_plan.m_plan;
+    }
+
+    void Execute (FFTplan& fft_plan)
+    {
+        if (!(fft_plan.m_stream == amrex::Gpu::gpuStream())) {
+            amrex::Gpu::streamSynchronize();
+        }
+
+        sycl::event r;
+        if (fft_plan.m_dir == direction::R2C) {
+            r = oneapi::mkl::dft::compute_forward(
+                *fft_plan.m_plan,
+                fft_plan.m_real_array,
+                reinterpret_cast<amrex::Real*>(fft_plan.m_complex_array));
+        } else {
+            r = oneapi::mkl::dft::compute_backward(
+                *fft_plan.m_plan,
+                reinterpret_cast<amrex::Real*>(fft_plan.m_complex_array),
```

Comment:
It looks to me like input & output types are distinct types in the API:
https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2024-1/compute-forward-typename-descriptor-type-typename.html

Is there a need to cast the complex pointers to `Real*`?

Reply:
Yes, it does not compile with `amrex::GpuComplex<Real>*`. Either `Real*` or `std::complex<Real>*` works.